### PR TITLE
Optimize compute unit limit request

### PR DESCRIPTION
**Pull request by codewithgun**

### CODE REVIEW

File: ts-client/src/dlmm/helpers/index.ts [hunk 1]
```diff
@@ -197,8 +198,70 @@ export async function chunkedGetMultipleAccountInfos(
...
+ * @returns The estimated compute unit usage with the buffer.
+ */
+export const getEstimatedComputeUnitUsageWithBuffer = async (
+  connection: Connection,
+  instructions: TransactionInstruction[],
+  feePayer: PublicKey,
+  buffer?: number
+) => {
+  if (!buffer) {
+    buffer = 0.1;
```
Comment (andrewsource147) [inline ts-client/src/dlmm/helpers/index.ts:214]: I think the surplus should be between 50k-200k 
the additional CU = (estimatedComputeUnitUsage * buffer).max(50k).min(200k)

File: ts-client/src/dlmm/helpers/index.ts [hunk 2]
```diff
@@ -197,8 +198,70 @@ export async function chunkedGetMultipleAccountInfos(
...
+    connection,
+    instructions,
+    feePayer,
+    buffer
+  ).catch((error) => {
+    console.error("Error::getEstimatedComputeUnitUsageWithBuffer", error);
+    if (fallbackUnit) {
+      return fallbackUnit;
+    }
+    throw error;
```
Comment (andrewsource147) [inline ts-client/src/dlmm/helpers/index.ts:263]: better to add fallBack to maxCU unit?

File: ts-client/src/dlmm/index.ts [hunk 3]
```diff
@@ -3072,14 +3150,50 @@ export class DLMM {
...
       .transaction();
 
+    const setCUIx = await getEstimatedComputeUnitIxWithBuffer(
+      this.program.provider.connection,
+      removeLiquidityTx.instructions,
+      user,
+      DEFAULT_REMOVE_LIQUIDITY_CU
+    );
+
+    removeLiquidityTx = await this.program.methods
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:3160]: feel a bit duplicated with the code above, is there a way to refactor that?

File: ts-client/src/dlmm/index.ts [hunk 4]
```diff
@@ -3522,6 +3636,36 @@ export class DLMM {
       .postInstructions(postInstructions)
       .transaction();
 
+    const setCUIx = await getEstimatedComputeUnitIxWithBuffer(
+      this.program.provider.connection,
+      swapTx.instructions,
+      user
+    );
+
+    swapTx = await this.program.methods
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:3645]: ditto^

File: ts-client/src/dlmm/index.ts [hunk 5]
```diff
@@ -3631,6 +3775,40 @@ export class DLMM {
...
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:3784]: ditto^

File: ts-client/src/dlmm/index.ts (hunk 3 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed, it actually create issue and causes test to fail

File: ts-client/src/dlmm/index.ts (hunk 4 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed

File: ts-client/src/dlmm/index.ts (hunk 5 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed

File: ts-client/src/dlmm/helpers/index.ts [hunk 6]
```diff
@@ -197,8 +198,70 @@ export async function chunkedGetMultipleAccountInfos(
...
+
+  const estimatedComputeUnitUsage = await getSimulationComputeUnits(
+    connection,
+    instructions,
+    feePayer,
+    []
+  );
+
+  const estimatedComputeUnitUsageWithBuffer = Math.min(
+    estimatedComputeUnitUsage * (1 + buffer),
```
[the remark this answers is not in the document]
Comment (codewithgun) [reply ts-client/src/dlmm/helpers/index.ts:229]: I think it's fine to still have it. It improves the landing rate.

File: ts-client/src/dlmm/helpers/index.ts (hunk 2 above)
Comment (codewithgun) [reply ts-client/src/dlmm/helpers/index.ts]: Fixed

File: ts-client/src/dlmm/helpers/index.ts (hunk 1 above)
Comment (codewithgun) [reply ts-client/src/dlmm/helpers/index.ts:217]: Fixed

File: ts-client/src/dlmm/helpers/index.ts [hunk 7]
```diff
@@ -197,8 +199,69 @@ export async function chunkedGetMultipleAccountInfos(
...
+
+  const estimatedComputeUnitUsage = await getSimulationComputeUnits(
+    connection,
+    instructions,
+    feePayer,
+    []
+  );
+
+  let extraComputeUnitBuffer = estimatedComputeUnitUsage * buffer;
+  if (extraComputeUnitBuffer > MAX_CU_BUFFER) {
```
Comment (00xSam) [inline ts-client/src/dlmm/helpers/index.ts:232]: this util should not limit to `add liquidity` function?

File: ts-client/src/dlmm/index.ts [hunk 8]
```diff
@@ -2165,19 +2171,31 @@ export class DLMM {
     const programMethod =
       this.program.methods.addLiquidityByStrategy(liquidityParams);
 
-    const createPositionTx = await programMethod
+    const createPositionIx = await programMethod
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:2174]: this is `addLiquidityIx`

File: ts-client/src/dlmm/index.ts [hunk 9]
```diff
@@ -2593,19 +2623,31 @@ export class DLMM {
...
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:2626]: ditto^

File: ts-client/src/dlmm/index.ts [hunk 10]
```diff
@@ -3070,32 +3123,51 @@ export class DLMM {
...
+
+    const setCUIx = await getEstimatedComputeUnitIxWithBuffer(
+      this.program.provider.connection,
+      instructions,
+      user
+    );
+
+    instructions.unshift(setCUIx);
 
     const { blockhash, lastValidBlockHeight } =
```
Comment (andrewsource147) [inline ts-client/src/dlmm/index.ts:3142]: that should be fetch inside the block if/else, right before return transaction

File: ts-client/src/dlmm/index.ts (hunk 8 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed

File: ts-client/src/dlmm/index.ts (hunk 9 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed

File: ts-client/src/dlmm/index.ts (hunk 10 above)
Comment (codewithgun) [reply ts-client/src/dlmm/index.ts]: Fixed

File: ts-client/src/dlmm/helpers/index.ts (hunk 7 above)
Comment (codewithgun) [reply ts-client/src/dlmm/helpers/index.ts:232]: It applies to all functions. It clamps only the extra CU buffer, not the estimated CU used.

**Review by andrewsource147 — APPROVED** [review]

**Review by 00xSam — APPROVED** [review]
Can u rebase?

**Pull request merged**